### ISSUE TITLE
No stack trace

### ISSUE THREAD

**Issue by createthis** [mocha-parallel/mocha-parallel-tests#202]
Hello,

I'm trying to set up `mocha-parallel-tests` with my test suite of about a thousand tests. I'm seeing random heisenbug style failures and no stack trace when the code fails:

```
> NODE_ENV=test node_modules/.bin/mocha-parallel-tests --exit --timeout 15000 --full-trace --require babel-core/register

[snip passing test groups]

Warning: React depends on requestAnimationFrame. Make sure that you load a polyfill in older browsers. https://fb.me/react-polyfills
before axio MockAdapter
after axio MockAdapter
              
                <Roles />
                  getItems
                    ✓ returns 1 (86ms)
                    1) "before each" hook for "returns 2"


npm ERR! Test failed.  See above for more details.
Jesses-MacBook-Pro:node jesse$ 
```

I also have a `test/mocha.opts` file:

```
--require should
--recursive
```

How can I debug this further? Thanks.

**Comment by 1999** [comment]
I'd say the issue here is probably that you're relying on some global variable available in all of your tests. Can you take a brief look at your tests and see id there's any?

**Comment by createthis** [comment]
1. Why would a global variable cause stack traces and test output to disappear?
1. How would I track down a global variable?

I `import` a single helper from all of my tests. I'm unclear on whether or not that would be considered a global variable in this context.

**Comment by 1999** [comment]
Yeah nah, I was mostly talking about the cases like this: https://github.com/mocha-parallel/mocha-parallel-tests/wiki/Limitations#global-variables-injected-by-reporters-are-not-fully-supported

Btw I just realized that showing stack traces is not something which mocha does as well: one of your "before" hooks fails and mocha reporter just shows that it failed. Can you try reducing the tests files number to a smaller amount and track the error?

**Comment by createthis** [comment]
I'm not sure what you mean. I added a `throw` to a `beforeEach` in a regular `mocha` test and I see a backtrace:

```
$ npm test -- -g pez

> NODE_ENV=test node_modules/.bin/mocha --exit --timeout 15000 --full-trace --require babel-core/register "-g" "pez"


  pez
    multipart form-data
      1) "before each" hook for "should throw an exception when boundary other than last is terminated"


  0 passing (7s)
  1 failing

  1) pez
       "before each" hook for "should throw an exception when boundary other than last is terminated":
     Error: oh no
      at Context.<anonymous> (/Users/jesse/Documents/node/test/services/pez.spec.js:10:11)
      at callFn (/Users/jesse/Documents/node/node_modules/mocha/lib/runnable.js:372:21)
      at Hook.Runnable.run (/Users/jesse/Documents/node/node_modules/mocha/lib/runnable.js:364:7)
      at next (/Users/jesse/Documents/node/node_modules/mocha/lib/runner.js:317:10)
      at Immediate.<anonymous> (/Users/jesse/Documents/node/node_modules/mocha/lib/runner.js:347:5)
      at runCallback (timers.js:810:20)
      at tryOnImmediate (timers.js:768:5)
      at processImmediate [as _immediateCallback] (timers.js:745:5)



npm ERR! Test failed.  See above for more details.
node jesse$ 
```

I would gladly show you the same output using `mocha-parallel-tests`, but #203 prevents me.

**Comment by 1999** [comment]
Gotcha, sorry for that. Let's jump to #203 first and then go back to this issue.

**Comment by createthis** [comment]
Ok, I think I've managed to reproduce the missing stack trace in a minimal git repo. It appears to be a race condition of some kind. Check it out: https://github.com/createthis/test_mocha_parallel_tests_traces

Reproduction steps:

```bash
git clone https://github.com/createthis/test_mocha_parallel_tests_traces.git
cd test_mocha_parallel_tests_traces
npm install
npm test # repeat this command until the stack trace disappears
```

Example output:

```
Jesses-MacBook-Pro:test_mocha_parallel_tests_traces jesse$ npm test

> test_mocha_parallel_tests@1.0.0 test /Users/jesse/test_mocha_parallel_tests_traces
> mocha-parallel-tests --exit --timeout 15000 --full-trace



  
    fake
      ✓ should be a fake test that passes



  
    other fake
      ✓ should be another fake test that passes



  
    boom
      nested describe
        1) "before each" hook
        ✓ should be a fake test that passes


npm ERR! Test failed.  See above for more details.
Jesses-MacBook-Pro:test_mocha_parallel_tests_traces jesse$
```

but note that you may have to run it a few times. Here's the very next run of the tests, in which I *do* get a stack trace:

```bash
Jesses-MacBook-Pro:test_mocha_parallel_tests_traces jesse$ npm test

> test_mocha_parallel_tests@1.0.0 test /Users/jesse/test_mocha_parallel_tests_traces
> mocha-parallel-tests --exit --timeout 15000 --full-trace



  
    other fake
      ✓ should be another fake test that passes



  
    fake
      ✓ should be a fake test that passes



  
    boom
      nested describe
        1) "before each" hook
        ✓ should be a fake test that passes




  3 passing (424ms)
  1 failing

  1) 
       boom
         "before each" hook:
     Error: boom!
      at Timeout.setTimeout [as _onTimeout] (test/boom.spec.js:4:13)



npm ERR! Test failed.  See above for more details.
Jesses-MacBook-Pro:test_mocha_parallel_tests_traces jesse$
```

HTH

**Comment by createthis** [comment]
I’m on an old 2013 MacBook, so I’m not surprised you couldn’t reproduce. My analysis of the problem is that it’s a race condition. It happens when the beforeEach throws an exception *after* the it test starts executing, but *before* it finishes. Only a theory. If you are on a faster machine, maybe try a 5 second await in the it block and a 1 second wait in the beforeEach?

Theoretically speaking, if there was such a race condition, where in the code might it be? I may read through and attempt to audit.

Mocha does the right thing in each case by displaying the stack trace. I’ve never been able to get mocha to not display the stack trace.

**Comment by 1999** [comment]
Not related to this race condition issue: I just realized that `--full-trace` is not supported at all, so the PR above fixes this. Could you please re-check this issue on a new version? `mocha-parallel-tests@2.0.6-alpha.2`

**Comment by createthis** [comment]
Yes, I'm seeing full traces now when the race condition doesn't get me:

```bash
  1) 
       boom
         "before each" hook:
     Error: boom!
      at Timeout.setTimeout [as _onTimeout] (/Users/jesse/test_mocha_parallel_tests_traces/test/boom.spec.js:4:13)
      at ontimeout (timers.js:498:11)
      at tryOnTimeout (timers.js:323:5)
      at Timer.listOnTimeout (timers.js:290:5)
```

still fails strangely when the race condition gets me though:

```bash
    boom
      nested describe
        1) "before each" hook
        ✓ should be a fake test that passes


npm ERR! Test failed.  See above for more details.
```

**Comment by createthis** [comment]
@1999 do a pull on `test_mocha_parallel_tests_traces` and try again. I added two more tests and it fails every time for me now.

**Comment by 1999** [comment]
@createthis you're a legend! I will look into this tomorrow.

**Comment by 1999** [comment]
Btw: it looks like the issue is not related to `--full-trace`: it's the `--exit` option that's so crucial for the test. If you remove the `--exit` everything will be working as expected, but I assume you use it for some reason?

**Comment by createthis** [comment]
Huh. Yes, you're right. It's `--exit`.  I was hoping this would provide me with a workaround so that I could see the backtraces related to my random heisenbug style test failures, but that appears not to be the case. See below.

## To answer your question:

I use `--exit` because my 1100+ test suite just hangs at the end and never quits. I have to kill it manually with `CTRL+c`.

I used to think this was just how `Mocha` worked by default, but now that I remove `--exit` from the `test_mocha_parallel_tests_traces` minimal test suite, I see it exits properly. Huh. I guess I learned something today.

My revised theory is that this hang at the end of my test suite is a side effect of importing either my `HapiJS` server for controller testing, or my `Sequelize` server. Perhaps a file or port is being left open? I'm not sure.

Anyway, yes, apparently I need `--exit` until I can track down the cause of the hang issue.

Using `mocha-parallel-tests` to execute my 1100+ test suite with no `--exit`, I never receive a test summary at the end, which means, unfortunately, I never receive backtraces.

## BTW
It appears that #204 is also caused by `--exit` as my test suite output stops scrolling horizontally when I remove `--exit` from `package.json`.

**Comment by 1999** [comment]
> My revised theory is that this hang at the end of my test suite is a side effect of importing either my HapiJS server for controller testing, or my Sequelize server. Perhaps a file or port is being left open? I'm not sure.

Definitely. IIRC you can try using https://github.com/mafintosh/why-is-node-running to see what's the problem. Anyway, it would be nice to fix the reporter issue because mocha somehow manages to show the stdout whereas mocha-parallel-tests fails here. I will ping you when the PR is merged.

**Comment by 1999** [comment]
Hey @createthis. First of all thanks for posting this issue. It may be one of the most useful for the whole year, so I really appreciate that. The thing about `--exit` that I mentioned above was an issue with the asynchronous nature of node.js master vs forked process IPC. These bugs are often difficult to find but hell yeah, now it's much more stable. Just in case you need it: `mocha-parallel-tests@2.0.6-alpha.3` has the fix which in your case will result in always visible stack traces.

Another thing I wanted to mention is your test. I spent a huge amount of time experimenting with your test cases from the repo and realized that the [beforeEach](https://github.com/createthis/test_mocha_parallel_tests_traces/blob/master/test/boom.spec.js#L2) example is actually buggy.

```javascript
beforeEach(async function A() {
  setTimeout(function B() {
    throw new Error('boom!');
  }, 1)
});
```

I added function names for simplicity. Although you define function A as asynchronous `beforeEach` will not wait for the function B to finish executing. This is because setTimeout itself is not asynchronous function and you're not using any `await` instruction. This means that `mocha` (or any other Javascript) cannot catch the exception that you throw in a try..catch block and it will be propagated to the top of the stack as an `uncaughtException`. This is not necessarily a problem, `mocha` knows how to catch these kinds of exceptions, just wanted to let you know about that so that you don't have expectations that `beforeEach` should fail. It can, but actually, it should not :)

**Comment by createthis** [comment]
@1999 confirmed, `alpha.3` appears to have solved the problem. In addition, it appears to have solved #204 as well. I'm seeing backtraces all the time now and my tests do not scroll horizontally. Thank you!
